fraud/service.go

```diff
@@ -115,30 +147,190 @@ func (f *service) processIncoming(
+}
 
-	return getAll(ctx, store, proofType)
+func (f *service) handleFraudMessageRequest(stream network.Stream) {
+	msg := new(pb.FraudMessage)
+	_, err := serde.Read(stream, msg)
+	if err != nil {
+		stream.Reset() //nolint:errcheck
+		log.Warn(err)
+		return
+	}
+
+	errg, ctx := errgroup.WithContext(context.Background())
+	mu := sync.Mutex{}
+	for _, p := range msg.RequestedProofType {
+		proofType := toProof(p)
+		errg.Go(func() error {
+			proofs, err := f.Get(ctx, proofType)
+			if err != nil {
+				if errors.Is(err, datastore.ErrNotFound) {
+					return nil
+				}
+				return err
+			}
+
+			for _, proof := range proofs {
+				bin, err := proof.MarshalBinary()
+				if err != nil {
+					log.Warn(err)
+					continue
+				}
+				resp := &pb.ProofResponse{Type: int32(proofType), Value: bin}
+				mu.Lock()
+				msg.Proofs = append(msg.Proofs, resp)
+				mu.Unlock()
+			}
+			return nil
+		})
+	}
+
+	if err = errg.Wait(); err != nil {
+		stream.Reset() //nolint:errcheck
+		log.Error(err)
+		return
+	}
+	_, err = serde.Write(stream, msg)
+	if err != nil {
+		log.Error(err)
+	}
+}
+
+func (f *service) Sync(ctx context.Context) {
+	go f.syncFraudProofs(ctx)
+}
+
+func (f *service) syncFraudProofs(ctx context.Context) {
+	log.Debug("start fetching Fraud Proofs")
+	sub, err := f.host.EventBus().Subscribe(&event.EvtPeerIdentificationCompleted{})
+	if err != nil {
+		log.Error(err)
+		return
+	}
+	defer sub.Close()
+	connStatus := event.EvtPeerIdentificationCompleted{}
+	for i := 0; i < fraudRequests; i++ {
+		select {
+		case <-ctx.Done():
+			return
+		case e := <-sub.Out():
+			connStatus = e.(event.EvtPeerIdentificationCompleted)
+		}
+
+		if connStatus.Peer == f.host.ID() {
+			continue
+		}
+		go func(pid peer.ID) {
+			log.Debug("requesting proofs from peer ", pid)
+			f.topicsLk.RLock()
+			proofTypes := make([]int32, 0, len(f.topics))
+			for proofType := range f.topics {
+				proofTypes = append(proofTypes, int32(proofType))
+			}
+			f.topicsLk.RUnlock()
+			respProofs, err := requestProofs(ctx, f.host, pid, proofTypes)
+			if err != nil {
+				log.Error(err)
+				return
+			}
+			if len(respProofs) == 0 {
+				log.Debug("proof was not found")
+				return
+			}
+			log.Debug("got fraud proofs from peer: ", connStatus.Peer)
+			wg := sync.WaitGroup{}
+			wg.Add(len(respProofs))
+			for _, data := range respProofs {
+				go func(data *pb.ProofResponse) {
+					defer wg.Done()
+					proof, err := Unmarshal(toProof(data.Type), data.Value)
+					if err != nil {
+						if !errors.Is(err, &errNoUnmarshaler{}) {
+							f.pubsub.BlacklistPeer(pid)
+						}
+						return
+					}
+					h, err := f.getter(ctx, proof.Height())
+					if err != nil {
+						log.Error(err)
+						return
+					}
+					err = proof.Validate(h)
+					if err != nil {
+						f.pubsub.BlacklistPeer(pid)
+						return
+					}
+
+					err = f.put(ctx, proof.Type(), proof.HeaderHash(), data.Value)
+					if err != nil {
+						log.Error(err)
+					}
```

Comment:
The validation and storing are redundant. This is why we do Publish with the local option, so the existing validation pipeline `processIncoming` can handle that.

Reply:
in case of local publication - from field contains our own peer id.

Reply:
In sync, we discussed a solution to use WithSecretKeyAndPeerID option with additional one liner change in the pubsub